### ISSUE TITLE
Utilization of the congestion window

### ISSUE THREAD

**Issue by larryliu2018** [lucas-clemente/quic-go#3585]
Hello, I test in the wifi scenario and find from qlog **that most Bytes in flight are under the congestion window, and the difference is large.** The number of acks in the test is also reduced because of the wifi link. **Does this mean that at some point the congestion window is not being fully utilized?** 

Also, in these cases I monitored the flow control window and found that it was unconstrained and wondered what might be done to mitigate the underutilized congestion window, which was not the case when the ACK ratio was 1:2. The number of acks in the test is also reduced because of the wifi link. Does this mean that at some point the congestion window is not being fully utilized? 
Also, in these cases, I monitored the flow control window and found that it was unconstrained and wondered what might be done to mitigate this problem, which was not the case when the ACK ratio was 1:2. 

The following is the case of ACK 1:2: 
![image](https://user-images.githubusercontent.com/35099991/194887069-aca5b283-aa3f-4b73-abfc-ca9b10ad44a7.png)

The following is the result of the test under wifi, ACK 1:100
![image](https://user-images.githubusercontent.com/35099991/194887040-5b2f2f15-0847-4880-a7c3-22f51e897317.png)

I have an idea I don't know if it's right. Bytes in flight have actually been received by the receiver, but due to the delay of ACK, it did not return ACK in time, resulting in many Bytes in flight showing unconfirmed status, so that many Bytes in flight in the qlog output figure are under the congestion window. And then it affects the pacing rate.

I really hope to get your guidance, thank you @marten-seemann

**Comment by marten-seemann** [comment]
> Does this mean that at some point the congestion window is not being fully utilized?

Likely. That's why we have #2607, #2877 and #3563.

**Comment by larryliu2018** [comment]
> > Does this mean that at some point the congestion window is not being fully utilized?
> 
> Likely. That's why we have #2607, #2877 and #3563.

Thank you very much for your quick reply. I have read the relevant issue you provided. The current state is caused by the fact that the linux kernel did not read the UDP packets in time, which led to such a situation, right?

**Comment by marten-seemann** [comment]
Not really. We already use batched reading. We need to use batched sending, so we can get those packets onto the wire faster.

**Comment by larryliu2018** [comment]
> Not really. We already use batched reading. We need to use batched sending, so we can get those packets onto the wire faster.

Thank you for your patient explanation. I probably understand that the underutilization of the congestion window shown above is caused by the fact that the UDP packets are not sent to the link in time. There was spare sending space, but the UDP packets cannot be sent faster, which leads to the problem of insufficient link utilization.
